go/pkg/tool/tool.go

```diff
@@ -326,18 +326,43 @@ func (c *Client) DownloadDirectory(ctx context.Context, rootDigest, path string)
 	return err
 }
 
+// UploadStats contains various metadata of a directory upload.
+type UploadStats struct {
+	rc.TreeStats
+	RootDigest           digest.Digest
+	TotalNumBlobs        int64
+	NumCacheMisses       int64
+	RealBytesUploaded    int64
+	LogicalBytesUploaded int64
+}
+
 // UploadDirectory uploads a directory from the specified path as a Merkle-tree to the remote cache.
-func (c *Client) UploadDirectory(ctx context.Context, path string) error {
+func (c *Client) UploadDirectory(ctx context.Context, path string) (*UploadStats, error) {
 	log.Infof("Computing Merkle tree rooted at %s", path)
 	root, blobs, stats, err := c.GrpcClient.ComputeMerkleTree(ctx, path, "", "", &command.InputSpec{Inputs: []string{"."}}, filemetadata.NewNoopCache())
 	if err != nil {
-		return err
+		return nil, err
 	}
 	log.Infof("Directory root digest: %v", root)
 	log.Infof("Directory stats: %d files, %d directories, %d symlinks, %d total bytes", stats.InputFiles, stats.InputDirectories, stats.InputSymlinks, stats.TotalInputBytes)
 	log.Infof("Uploading directory %v rooted at %s to CAS.", root, path)
-	_, _, err = c.GrpcClient.UploadIfMissing(ctx, blobs...)
-	return err
+	missing, n, err := c.GrpcClient.UploadIfMissing(ctx, blobs...)
+	if err != nil {
+		return nil, err
+	}
+	var sumMissingBytes int64
+	for _, d := range missing {
+		sumMissingBytes += d.Size
+	}
+	us := &UploadStats{
+		TreeStats:            *stats,
+		RootDigest:           root,
+		TotalNumBlobs:        int64(len(blobs)),
+		NumCacheMisses:       int64(len(missing)),
+		RealBytesUploaded:    n,
```

Comment:
Would you mind using terminology like CacheHits / CacheMisses for sizes too? Real and Logical can be more obscure to folks unfamiliar with the system.

Reply:
Oh, they're not the same thing -- the difference between real and logical is compression only -- TotalInputBytes is the total size of the entire directory + metadata protos. Maybe it would be better if I omit the real/logical distinction altogether and just report the TotalBytesUploaded instead?

Reply:
Ah! Maybe use "CompressedBytesUploaded" then instead of RealBytesUploaded?  I think it's really useful data to have! Then I'd advocate for "CacheMissesBytes" for LogicalBytesUploaded. I don't feel "BytesCacheMisses" sounds good but it'd be nice to be coherent with "NumCacheMisses".

Reply:
Sure, done. But I worry that CompressedBytesUploaded might sound misleading, because it will only be different from CacheMissesBytes if compression is enabled.

Reply:
I might suggest "BytesTransferred" for actual bytes moved over the wire, "BytesRequested" for total bytes provided in the request, and "BytesProcessed" for actual bytes processed locally (may be less than BytesRequested if processing was interrupted).

Reply:
That's a great idea.

Reply:
I didn't get it, sorry -- CompressedBytesUploaded->BytesTransferred, and what should I rename CacheMissesBytes to? Also, if any error happened I prefer not to return any stats at all, because then it's not clear from the tool output that an error occurred (lmk if that's wrong and we want to print partial stats).

Reply:
Good question. The output should be explicit about errors. I agree that in this case, an non-zero exit code with no output is sufficient since the errors are being logged to stderr. The "BytesProcessed" is useful internally when it's combined from multiple requests, which is not in scope for this tool.

Reply:
Hmm, I was thinking about it from the perspective of an interactive use-case, where the user can spot-check stderr. If we're going to support scripting workflows, we'd need to make sure that both the exit code and an error field are set since users might use either or both. Also, the error message now needs to be printed on both streams when `--json` is provided, which might call for some utility functions.

Reply:
Okay, renamed the field, leaving out BytesProcessed for now, added an Error field, and printed the JSON unconditionally. PTAL.

Reply:
lgtm